File: tree/LCA.cpp
/**
 * find LCA of two tree nodes
 */

#include <iostream>

using namespace std;

class treeNode {
 public:
  treeNode *left;
  treeNode *right;
  int data;
  treeNode(int item) {left=NULL; right=NULL;data = item;};
};

class tree {
 public:
  treeNode *root;
  treeNode *LCA(treeNode *a, treeNode *b);
  treeNode *LCA(treeNode *root, treeNode *a, treeNode *b);
  bool isAChildOfB(treeNode *a, treeNode *b);
};

bool tree::isAChildOfB(treeNode *a, treeNode *b) {
  if (b == NULL) return false;
  if (a == b) return true;
  return isAChildOfB(a, b->left) || isAChildOfB(a, b->right);
}

treeNode *tree::LCA(treeNode *a, treeNode *b)  {
  return isAChildOfB(a, b)? b : (isAChildOfB(b, a)? a : LCA(root, a, b));
}

treeNode *tree::LCA(treeNode *root, treeNode *a, treeNode *b)  {
  if (root == NULL) return NULL;

  if (root->left == a || root->right == a || root->left == b || root->right == b)
    return root;

  treeNode *left = LCA(root->left, a, b);
  treeNode *right = LCA(root->right, a, b);

  return (left != NULL && right != NULL)? root : ((left != NULL)? left : right);
}

treeNode *tree::LCA(treeNode *root, treeNode *a, treeNode *b)  {
    if (!root) return NULL;
    if (root == a || root == b) return root;
    treeNode *left = LCA(root->left, a, b);
    treeNode *right = LCA(root->right, a, b);
    return (left && right)? root : (left? left : right);
}

int main() {
  tree *tree1 = new tree;
  treeNode *t1 = new treeNode(1);
  treeNode *t2 = new treeNode(2);
  treeNode *t3 = new treeNode(3);
  treeNode *t4 = new treeNode(4);
  treeNode *t5 = new treeNode(5);
  treeNode *t6 = new treeNode(6);
  tree1->root = t1;
  t1->left = t2;
  t1->right = t3;
  t2->left = t4;
  t4->left = t5;
  t2->right = t6;
  treeNode *t0 = tree1->LCA(t2, t6);
  cout<<t0->data<<endl;
}

File: tree/LCA.java
import java.util.*;

public class LCA {
    // Solution 1: if BST, just use bst search 
    public TreeNode<Integer> LCA_BST(TreeNode<Integer> root,
            TreeNode<Integer> n1, TreeNode<Integer> n2) {
        if (root == null) return null;
        int smaller = Math.min(n1.data,n2.data);
        int bigger  = Math.max(n1.data,n2.data);
        if (root.data < smaller) 
            return LCA_BST(root.right, n1, n2);
        else if (root.data > bigger)
            return LCA_BST(root.left, n1, n2);
        else 
            return root;
    }

    // Solution 2: for general tree
    // Need to check n1/n2 relationship first
    public TreeNode<Integer> getLCA(TreeNode<Integer> root,
            TreeNode<Integer> n1, TreeNode<Integer> n2) {
        if (root == n1 || root == n2) return root;
        if (isAChildOfB(n1, n2)) return n2;
        if (isAChildOfB(n2, n1)) return n1;
        return LCA(root, n1, n2);
    }

    public boolean isAChildOfB(TreeNode<Integer> n1, TreeNode<Integer> n2) {
        if (n2 == null) return false;
        if (n1 == n2) return true;
        return isAChildOfB(n1, n2.left) || isAChildOfB(n1, n2.right);
    }

    public TreeNode<Integer> LCA(TreeNode<Integer> root,
            TreeNode<Integer> n1, TreeNode<Integer> n2) {
        if (root == null) return null;
        if (root.left == n1 || root.left == n2 
                || root.right == n1 || root.right == n2) {
            return root;
        }
        TreeNode<Integer> left = LCA(root.left, n1, n2);
        TreeNode<Integer> right = LCA(root.right, n1, n2);
        if (left != null && right != null)
            return root;
        return left != null? left : right;
    }

    // Solution 3: find path for n1, n2, 
    // and get Longest common path
    public boolean LCA_PATH_Helper(TreeNode<Integer> root,
            TreeNode<Integer> n,
            List<TreeNode<Integer>> l) {
        if (root == null) return false;
        if ((root.data == n.data) || LCA_PATH_Helper(root.left,n,l) || LCA_PATH_Helper(root.right,n,l)) {
            l.add(root);
            return true;
        }
        return false;
    }

    public TreeNode<Integer> LCA_PATH(TreeNode<Integer> root,
            TreeNode<Integer> n1, TreeNode<Integer> n2) {
        if (root == null) return null;
        if (n1.left == n2 || n1.right == n2) return n1;
        if (n2.left == n1 || n2.right == n2) return n2;
        List<TreeNode<Integer>> l1 = new ArrayList<TreeNode<Integer>>();
        List<TreeNode<Integer>> l2 = new ArrayList<TreeNode<Integer>>();
        LCA_PATH_Helper(root, n1, l1);
        LCA_PATH_Helper(root, n2, l2);
        int i = l1.size()-1;
        int j = l2.size()-1;
        TreeNode<Integer> lca = null;
        while (i>=0 && j>=0) {
            if (l1.get(i) != l2.get(j)) break;
            lca = l1.get(i);
            i--;
            j--;
        }
        return lca;
    }

    public void testLCA() {
        Tree<Integer> tree1 = new Tree<Integer>();
        TreeNode<Integer> n4 = new TreeNode<Integer>(4);
        TreeNode<Integer> n2 = new TreeNode<Integer>(2);
        TreeNode<Integer> n5 = new TreeNode<Integer>(5);
        TreeNode<Integer> n1 = new TreeNode<Integer>(1);
        TreeNode<Integer> n3 = new TreeNode<Integer>(3);
        TreeNode<Integer> n0 = new TreeNode<Integer>(0);
        tree1.root = n4;
        n4.left = n2;
        n4.right = n5;
        n2.left = n1;
        n2.right = n3;
        n1.left = n0;
        tree1.printByLevel();
        System.out.println(LCA_PATH(tree1.root, n1, n2).data);
        System.out.println(LCA_BST(tree1.root, n1, n2).data);
        System.out.println(getLCA(tree1.root, n0, n2).data);
    }

    public static void main(String[] args) {
        LCA LCA = new LCA();
        LCA.testLCA();
    }
}
